jmclient/jmclient/wallet_rpc.py

```diff
@@ -0,0 +1,674 @@
+            return make_jmwalletd_response(request)
+
+        @app.route('/wallet/<string:walletname>/maker/stop', methods=['GET'])
+        def stop_maker(self, request, walletname):
+            self.check_cookie(request)
+            if not self.wallet_service:
+                raise NoWalletFound()
+            if not self.services["maker"]:
+                raise ServiceNotStarted()
+            self.services["maker"].stopService()
+            return make_jmwalletd_response(request)
+
+        @app.route('/wallet/<string:walletname>/lock', methods=['GET'])
+        def lockwallet(self, request, walletname):
+            print_req(request)
+            self.check_cookie(request)
+            if not self.wallet_service:
+                jlog.warn("Called lock, but no wallet loaded")
+                # we could raise NoWalletFound here, but is
+                # easier for clients if they can gracefully call
+                # lock multiple times:
+                already_locked = True
+            else:
+                self.wallet_service.stopService()
+                self.cookie = None
+                self.wss_factory.valid_token = None
+                self.wallet_service = None
+                already_locked = False
+            return make_jmwalletd_response(request, walletname=walletname,
+                                           already_locked=already_locked)
+
+        @app.route('/wallet/create', methods=["POST"])
+        def createwallet(self, request):
+            print_req(request)
+            # we only handle one wallet at a time;
+            # if there is a currently unlocked wallet,
+            # refuse to process the request:
+            if self.wallet_service:
+                raise WalletAlreadyUnlocked()
+            request_data = self.get_POST_body(request,
+                            ["walletname", "password", "wallettype"])
+            wallettype = request_data["wallettype"]
+            if wallettype == "sw":
+                wallet_cls = SegwitWallet
+            elif wallettype == "sw-legacy":
+                wallet_cls = SegwitLegacyWallet
+            elif wallettype == "sw-fb":
+                wallet_cls = SegwitWalletFidelityBonds
+            else:
+                raise InvalidRequestFormat()
+            # use the config's data location combined with the json
+            # data to construct the wallet path:
+            wallet_root_path = os.path.join(jm_single().datadir, "wallets")
+            wallet_name = os.path.join(wallet_root_path,
+                                       request_data["walletname"])
+            try:
+                wallet = create_wallet(wallet_name,
+                                       request_data["password"].encode("ascii"),
+                                       4, wallet_cls=wallet_cls)
+                seedphrase_help_string = wallet_showseed(wallet)
+            except StorageError as e:
+                raise NotAuthorized(repr(e))
```

Comment:
Yes, good call. I've checked again what the structure of Exception types is, in `storage.py`, and:

Currently there is `StorageError` `RetryableStorageError(StorageError)`, and `StoragePasswordError(RetryableStorageError)` as derived cases, the second type delineates when the file *can* be opened, it  just isn't authorized right now (wrong password; read only; .lock exists).

That means that in the `createwallet` operation if we get a StorageError it can only mean (a) a .lock file exists with this name, or (b) "file already exists" (or perhaps not writeable?, but that means way bigger problems..), as the other cases don't apply. So I'll add a prior `RetryableStorageError` as a catch for the `.lock`, and a new response code type. Unlikely case but not super-unlikely. And then the `StorageError` after that for "file exists" (as in the actual .jmdat file). And I will put 409 Already Exists for that catch,


While in the `unlockwallet` operation if it's not a `StoragePasswordError` or a `RetryableStorageError` it must be the case of "file does not exist".

So in the `unlockwallet` case I think we already have it correctly covered with the three exceptions in succession on lines 559-566.

Reply:
Done in https://github.com/JoinMarket-Org/joinmarket-clientserver/pull/996/commits/28f37e5664d9e8742b0814be45ce0b7fefbbf73d

Note I added to the method descriptions the fact that 'this operation can't be performed when another wallet is already unlocked', those are the 401 responses, but the error message should make it clear.